In [2]:
import random
import copy

def create (alphabet, N=100): # crea población
    return [random.sample(alphabet, len(alphabet)) for individuo in range(N)]

def sort_pop (pop, fit): # ordena población por fitness debe devolver la poblacion ordenada y el fitness ordenado tmb (de cada individuo) utilizar el zip para generar tuplas de fitness - individuo
    fits = [fit(individuo) for individuo in pop]
    fits, population = zip(*sorted(zip(fits, pop)))
    return fits, population

def select (pop, T): # devuelve un individuo seleccionado por torneo, devuelve una copia para evitar efectos laterales
    seleccionados = random.sample(pop, T) #los 3 aleatorios
    return copy.copy(pop[min([pop.index(x) for x in seleccionados])])

def crossover (ind1, ind2, pcross):
    mascara = [random.choice([0, 1]) for i in range(len(ind1))] # 0 -> no se debe intercambiar, 1 sí

    copia_ind1 = [None if mascara[i] == 0 and random.random() < pcross else elem for i, elem in enumerate(ind1)]
    copia_ind2 = [None if mascara[i] == 0 and random.random() < pcross else elem for i, elem in enumerate(ind2)]

    for i, numero in enumerate(copia_ind1):
      if numero == None:
        for j, dato in enumerate(ind2):
          if dato not in copia_ind1:
            copia_ind1[i] = dato
            break

    for i, numero in enumerate(copia_ind2):
          if numero == None:
            for j, dato in enumerate(ind1):
              if dato not in copia_ind2:
                copia_ind2[i] = dato
                break

    return copia_ind1, copia_ind2


def create (alphabet, N=100): # crea población
    return [random.sample(alphabet, len(alphabet)) for individuo in range(N)]

def sort_pop (pop, fit): # ordena población por fitness debe devolver la poblacion ordenada y el fitness ordenado tmb (de cada individuo) utilizar el zip para generar tuplas de fitness - individuo
    fits = [fit(individuo) for individuo in pop]
    fits, population = zip(*sorted(zip(fits, pop)))
    return fits, population

def select (pop, T): # devuelve un individuo seleccionado por torneo, devuelve una copia para evitar efectos laterales
    seleccionados = random.sample(pop, T) #los 3 aleatorios
    return copy.copy(pop[min([pop.index(x) for x in seleccionados])])

def crossover (ind1, ind2, pcross):
    mascara = [random.choice([0, 1]) for i in range(len(ind1))] # 0 -> no se debe intercambiar, 1 sí

    copia_ind1 = [None if mascara[i] == 0 and random.random() < pcross else elem for i, elem in enumerate(ind1)]
    copia_ind2 = [None if mascara[i] == 0 and random.random() < pcross else elem for i, elem in enumerate(ind2)]

    for i, numero in enumerate(copia_ind1):
      if numero == None:
        for j, dato in enumerate(ind2):
          if dato not in copia_ind1:
            copia_ind1[i] = dato
            break

    for i, numero in enumerate(copia_ind2):
          if numero == None:
            for j, dato in enumerate(ind1):
              if dato not in copia_ind2:
                copia_ind2[i] = dato
                break

    return copia_ind1, copia_ind2


def mutate (ind, pmut): # la mutación consistirá en intercambiar elementos
    probs = [random.random() for prob in range(len(ind))]
    nuevo_elem = copy.copy(ind)
    for i, probability in enumerate(probs):
      if probability < pmut:
        elegido = random.randint(0, len(nuevo_elem) - 1)
        aux = nuevo_elem[i]
        nuevo_elem[i] = nuevo_elem[elegido]
        nuevo_elem[elegido] = aux
    return nuevo_elem

def evolve (pop, fit, pmut, pcross=0.7, ngen=100, T=2, trace=0, elitism=False, maximize=False):
    next_pop = []
    for n in range(ngen):
        fitness, pop = sort_pop(pop, fit)
    if elitism:
        next_pop.append(pop[0])
        
    for i in range(int(len(pop)/2)):
        
        individuo_elegido1 = select(pop, T)
        individuo_elegido2 = select(pop, T)
        individuo_elegido1, individuo_elegido2 = crossover(individuo_elegido1, individuo_elegido1, pcross)

        individuo_elegido1 = mutate(individuo_elegido1, pmut)
        
        individuo_elegido2 = mutate(individuo_elegido2, pmut)
        
        next_pop.append(individuo_elegido1)

        next_pop.append(individuo_elegido2)

    pop = copy.deepcopy(next_pop)
    next_pop = [] 

    _, pop = sort_pop(pop, fit)
    return pop
        
        

def evolve (pop, fit, pmut, pcross=0.7, ngen=100, T=2, trace=0, elitism=False, maximize=False):
  next_pop = []
  for _ in range(ngen):
    fitness, pop = sort_pop(pop, fit)
    if elitism:
      next_pop.append(pop[0])
      
    for i in range(int(len(pop)/2)):
      individuo_elegido1 = select(pop, T)
      individuo_elegido2 = select(pop, T)
      individuo_elegido1, individuo_elegido2 = crossover(individuo_elegido1, individuo_elegido1, pcross)

      individuo_elegido1 = mutate(individuo_elegido1, pmut)
      
      individuo_elegido2 = mutate(individuo_elegido2, pmut)
      
      next_pop.append(individuo_elegido1)

      next_pop.append(individuo_elegido2)

    pop = copy.deepcopy(next_pop)
    next_pop = [] 

  _, pop = sort_pop(pop, fit)
  return pop
        
        

In [3]:
import networkx as nx

amazon_G = nx.read_graphml('./amazon_graph.graphml')

# Minimizar la distancia entre los nodos dentro de una comunidad
# Maximizar la distancia entre los nodos de disintas comunidades

In [4]:
import numpy as np

def create_individuo(nodes_of_graph, adj_matrix, transformation):
    individuo =  [None] * len(nodes_of_graph)
    for i in range(len(nodes_of_graph)): #el individuo creado con el tipo locus tiene la longitud de los nodos

        
        nodo_vecino = random.choice(range(len(adj_matrix[i]))) #la matriz de adyacencia tiene los nodos vecinos, los nodos vecinos hay que convertirlos a numeros
        while adj_matrix[i][nodo_vecino] == 0 and nodo_vecino != i:
            nodo_vecino = random.choice(range(len(adj_matrix[nodo_vecino])))
        
        
        keys = list(transformation.keys())
        values = list(transformation.values())
        individuo[i] = keys[values.index(nodo_vecino)] # en el gen se pone el aleatorio de los vecinos pero transformado en el indice que le corresponde 
    return individuo

def create(nodes_of_graph, adj_matrix, transformation, N=100): # crea población, la transformacion debe de ser del tipo { numero_de_nodo_real : indice_del_array}, donde el indice del array va de 0 a len(nodos) - 1
    new_population = []
    for individuo in range(N):
        new_population.append(create_individuo(nodes_of_graph, adj_matrix, transformation))
    
    return new_population


In [5]:
conversiones = {}
for i, nodo in enumerate(amazon_G.nodes):
    conversiones[nodo] = i
print(conversiones)

{'75780': 0, '317458': 1, '186386': 2, '419862': 3, '176159': 4, '204831': 5, '104481': 6, '57383': 7, '401449': 8, '299055': 9, '176184': 10, '542780': 11, '366654': 12, '124992': 13, '16457': 14, '196682': 15, '546890': 16, '546891': 17, '34895': 18, '321619': 19, '112724': 20, '276567': 21, '55386': 22, '374874': 23, '55388': 24, '221281': 25, '432229': 26, '104553': 27, '383087': 28, '335987': 29, '448649': 30, '522378': 31, '59532': 32, '491667': 33, '518306': 34, '309411': 35, '182435': 36, '204966': 37, '532649': 38, '544939': 39, '542899': 40, '172218': 41, '270524': 42, '250046': 43, '204991': 44, '231620': 45, '497867': 46, '186575': 47, '346322': 48, '399572': 49, '276695': 50, '530647': 51, '18650': 52, '252125': 53, '59613': 54, '127199': 55, '379102': 56, '159972': 57, '366824': 58, '116969': 59, '202986': 60, '430314': 61, '350442': 62, '4328': 63, '458993': 64, '30962': 65, '35067': 66, '57600': 67, '217346': 68, '295177': 69, '49424': 70, '278': 71, '123159': 72, '3218

In [6]:
print(amazon_G.edges)

[('75780', '182923'), ('75780', '423864'), ('75780', '458993'), ('317458', '480479'), ('317458', '500861'), ('317458', '500865'), ('186386', '299055'), ('186386', '394703'), ('186386', '414657'), ('186386', '521655'), ('419862', '33251'), ('419862', '98081'), ('419862', '152407'), ('419862', '458993'), ('419862', '496201'), ('176159', '19509'), ('176159', '30962'), ('176159', '190765'), ('176159', '230788'), ('176159', '373756'), ('176159', '404948'), ('176159', '428737'), ('204831', '28513'), ('204831', '159972'), ('204831', '189286'), ('204831', '313801'), ('204831', '342330'), ('204831', '379184'), ('104481', '394886'), ('57383', '27890'), ('57383', '165463'), ('57383', '219781'), ('57383', '238491'), ('57383', '335987'), ('57383', '389895'), ('57383', '458993'), ('57383', '483076'), ('57383', '538129'), ('401449', '187924'), ('401449', '198973'), ('401449', '470620'), ('299055', '154202'), ('176184', '256445'), ('542780', '9762'), ('542780', '16457'), ('542780', '214555'), ('542780

In [7]:
A = nx.to_numpy_array(amazon_G)
print(A[0][0])


0.0


In [8]:

pop = create(amazon_G.nodes, A, conversiones)

In [9]:
print(conversiones)

{'75780': 0, '317458': 1, '186386': 2, '419862': 3, '176159': 4, '204831': 5, '104481': 6, '57383': 7, '401449': 8, '299055': 9, '176184': 10, '542780': 11, '366654': 12, '124992': 13, '16457': 14, '196682': 15, '546890': 16, '546891': 17, '34895': 18, '321619': 19, '112724': 20, '276567': 21, '55386': 22, '374874': 23, '55388': 24, '221281': 25, '432229': 26, '104553': 27, '383087': 28, '335987': 29, '448649': 30, '522378': 31, '59532': 32, '491667': 33, '518306': 34, '309411': 35, '182435': 36, '204966': 37, '532649': 38, '544939': 39, '542899': 40, '172218': 41, '270524': 42, '250046': 43, '204991': 44, '231620': 45, '497867': 46, '186575': 47, '346322': 48, '399572': 49, '276695': 50, '530647': 51, '18650': 52, '252125': 53, '59613': 54, '127199': 55, '379102': 56, '159972': 57, '366824': 58, '116969': 59, '202986': 60, '430314': 61, '350442': 62, '4328': 63, '458993': 64, '30962': 65, '35067': 66, '57600': 67, '217346': 68, '295177': 69, '49424': 70, '278': 71, '123159': 72, '3218

In [10]:
print(pop[0])

['75780', '317458', '521655', '496201', '190765', '379184', '394886', '335987', '198973', '154202', '176184', '9762', '366654', '460073', '16457', '216152', '546890', '126430', '306636', '475666', '114378', '276567', '55386', '374874', '165463', '229209', '141635', '104553', '289953', '27890', '448649', '64879', '59532', '374874', '518306', '127262', '186575', '84839', '532649', '214555', '216152', '172218', '270524', '296111', '256660', '280255', '326849', '186575', '435398', '399572', '482622', '29354', '359086', '108259', '59613', '208714', '49424', '342330', '318538', '34895', '502212', '30962', '361524', '123953', '301761', '19509', '59613', '57600', '217346', '127928', '30093', '79870', '137050', '321818', '321821', '377247', '500861', '370979', '313954', '285798', '154509', '330328', '546890', '546890', '190765', '185312', '379184', '543516', '141618', '45362', '149809', '313801', '198973', '383731', '141635', '447224', '275861', '279830', '14686', '403034', '340331', '59759', '

In [11]:
def convert_individuo(individuo_raw, conversiones):
    new_individuo = [None] * len(individuo_raw)
    for i in range(len(individuo_raw)):
        new_individuo[i] = conversiones.get(str(individuo_raw[i]))
    
    return new_individuo
def unconvert_individuo(individuo, conversiones):
    individuo_raw =  [None] * len(individuo)
    keys = list(conversiones.keys())
    values = list(conversiones.values())
    for i in range(len(individuo)):
        individuo_raw[i] = keys[values.index(individuo[i])]
    return individuo_raw

def get_comunidades(individuo):

    G = nx.Graph()

    for i in range(len(individuo)):
        G.add_edge(i, individuo[i])

    comunidades = list(nx.connected_components(G))
    print(comunidades)
    return comunidades

def get_cs(aristas_of_min_one_node_in_comunity, nodos_de_la_comunidad):
    cs = 0
    for aristas in aristas_of_min_one_node_in_comunity:
        if aristas[0] in nodos_de_la_comunidad and aristas[1] not in nodos_de_la_comunidad:
            cs += 1
            
        #elif aristas[1] in nodos_de_la_comunidad and aristas[0] not in nodos_de_la_comunidad:
        #    cs += 1
    return cs

def get_ms(aristas_of_min_one_node_in_comunity, nodos_de_la_comunidad):
    ms = 0
    for aristas in aristas_of_min_one_node_in_comunity:
        if aristas[0] in nodos_de_la_comunidad and aristas[1] in nodos_de_la_comunidad:
            ms += 1
            
        
    return ms


In [12]:
print(amazon_G.edges(str(75780)))

[('75780', '182923'), ('75780', '423864'), ('75780', '458993')]


In [13]:
def fit_conductance(grafo,individuo,conversiones):
    conductance = 0
    individuo_converted = convert_individuo(individuo, conversiones)
    comunidades = get_comunidades(individuo_converted)
    for comunidad in comunidades:
        aristas_of_min_one_node_in_comunity = []
        comunidad_desconvertida = unconvert_individuo(list(comunidad), conversiones)
        for nodo_of_comunity in comunidad_desconvertida:

            aristas_of_min_one_node_in_comunity.append(list(grafo.edges(str(nodo_of_comunity)))[0])
        #aqui ya tenemos las aristas de los nodos para poder sacar el sm
        cs_community = get_cs(aristas_of_min_one_node_in_comunity, comunidad_desconvertida)
        ms_community = get_ms(aristas_of_min_one_node_in_comunity, comunidad_desconvertida)
        conductance += cs_community / (2*ms_community + cs_community)
        #
    
    return conductance
        
        
def fit_expansion(grafo,individuo,conversiones):
    expansion = 0
    individuo_converted = convert_individuo(individuo, conversiones)
    comunidades = get_comunidades(individuo_converted)
    for comunidad in comunidades:
        aristas_of_min_one_node_in_comunity = []
        comunidad_desconvertida = unconvert_individuo(list(comunidad), conversiones)
        for nodo_of_comunity in comunidad_desconvertida:

            aristas_of_min_one_node_in_comunity.append(list(grafo.edges(str(nodo_of_comunity)))[0])
        #aqui ya tenemos las aristas de los nodos para poder sacar el sm
        cs_community = get_cs(aristas_of_min_one_node_in_comunity, comunidad_desconvertida)
        ns_community = len(comunidad_desconvertida)
        expansion += cs_community / ns_community
        #
    
    return expansion


def fit_average_odf(grafo,individuo,conversiones):
    average_odf = 0
    individuo_converted = convert_individuo(individuo, conversiones)
    comunidades = get_comunidades(individuo_converted)
    for comunidad in comunidades:
        comunidad_desconvertida = unconvert_individuo(list(comunidad), conversiones)
        sum_nodos = 0
        for nodo_of_comunity in comunidad_desconvertida:
            aristas_del_nodo = list(grafo.edges(str(nodo_of_comunity)))[0]
            
            sum_nodos += get_cs(aristas_del_nodo,nodo_of_comunity) / len(aristas_del_nodo)
            #aristas_of_min_one_node_in_comunity.append(list(grafo.edges(str(nodo_of_comunity)))[0])
        average_odf += sum_nodos / len(comunidad)
        
        #aqui ya tenemos las aristas de los nodos para poder sacar el sm
        
        #
    
    return average_odf

def fit_Q(grafo,individuo,conversiones):
    Q = 0
    individuo_converted = convert_individuo(individuo, conversiones)
    comunidades = get_comunidades(individuo_converted)
    for comunidad in comunidades:
        aristas_of_min_one_node_in_comunity = []
        comunidad_desconvertida = unconvert_individuo(list(comunidad), conversiones)
        n_aristas = 0
        for nodo_of_comunity in comunidad_desconvertida:
            n_aristas += len(list(grafo.edges(str(nodo_of_comunity)))[0])
            aristas_of_min_one_node_in_comunity.append(list(grafo.edges(str(nodo_of_comunity)))[0])
        
        ms_community = get_ms(aristas_of_min_one_node_in_comunity, comunidad_desconvertida)
        cs_community = get_cs(aristas_of_min_one_node_in_comunity, comunidad_desconvertida)
        
        Q += ((ms_community/n_aristas) - (((ms_community + cs_community)/(2*n_aristas))**2))
    
    return Q

#def fit(fit1, fit2, grafo, individuo, conversiones):
#    return fit1(grafo,individuo,conversiones) + fit2(grafo,individuo,conversiones)

#Esto no se si es asi por que al final lo que estas haciendo es convertir un problema multivariable en una combinacion muy simple, sin estar en la misma escala si quiera
        
print(fit_conductance(amazon_G, pop[1], conversiones))
print(fit_expansion(amazon_G, pop[1], conversiones))
print(fit_average_odf(amazon_G, pop[1], conversiones))
print(fit_Q(amazon_G, pop[1], conversiones))

[{0, 462}, {73, 1, 301}, {2}, {3, 102, 205, 338, 436, 126}, {4, 100, 198, 453, 456, 365, 142, 178, 115, 342, 407, 378, 315, 93}, {121, 5, 39, 200, 210, 86, 151, 472, 57, 91, 445}, {6}, {64, 224, 7, 202, 441, 330, 111, 463, 310, 153, 218, 379}, {8, 395, 92, 381}, {9, 147}, {10, 117}, {386, 40, 11, 398, 14}, {258, 12, 244, 103}, {56, 321, 13, 70}, {384, 370, 268, 15}, {288, 289, 418, 367, 16, 369, 17, 82, 20, 83, 371, 406, 253}, {458, 18, 473, 90}, {424, 19, 195}, {216, 154, 21, 430}, {104, 24, 22, 215}, {33, 23}, {25, 355}, {26, 309}, {27, 471}, {391, 265, 276, 279, 155, 28, 31, 35, 438, 58, 60, 75, 206, 335, 337, 469, 359, 110, 245, 120, 254, 383}, {325, 199, 304, 252, 412, 29}, {465, 260, 30}, {32}, {34, 197}, {36, 357, 394, 47, 468, 345, 444, 190}, {37, 454, 466, 211, 437, 119}, {256, 323, 38, 294, 295, 300, 237, 76, 271, 239, 273, 272, 339, 274, 275, 343, 344, 125}, {449, 101, 41, 459, 303, 470, 376}, {42, 419}, {98, 43, 332, 173, 281, 188}, {132, 389, 136, 396, 269, 283, 284, 156, 

In [ ]:
def select (pop, fits, T): # devuelve un individuo seleccionado por torneo, devuelve una copia para evitar efectos laterales
    seleccionados = random.sample(pop, T) #los 3 aleatorios
    es_dominante = True
    for i, indiv in enumerate(seleccionados):
        for j, indiv2 in enumerate(seleccionados): 
            if i == j:
                continue
            
            if fits[i][0] > fits[j][0] and fits[i][1] > fits[j][1]:
                es_dominante = (True and es_dominante)
            else:
                es_dominante = (False and es_dominante)
        if es_dominante:
            return indiv
    return random.choice(seleccionados)

In [ ]:
def normalizar_valores(fits):
    primeros_valores = [tupla[0] for tupla in fits]
    
    min_valor1 = min(primeros_valores)
    max_valor1 = max(primeros_valores)
    rango1 = max_valor1 - min_valor1

    primeros_valores = [tupla[1] for tupla in fits]
    min_valor2 = min(segundos_valores)
    max_valor2 = max(segundos_valores)
    rango2 = max_valor2 - min_valor2
    
    lista_normalizada = [((tupla[0] - min_valor1) / rango1, (tupla[1] - min_valor2) / rango2) for tupla in fits]
    
    return fits_normalizados



def sort_pop (pop, fit1, fit2, grafo, conversiones): # ordena población por fitness debe devolver la poblacion ordenada y el fitness ordenado tmb (de cada individuo) utilizar el zip para generar tuplas de fitness - individuo
    fits = [(fit1(grafo, individuo, conversiones), fit2(grafo, individuo, conversiones)) for individuo in pop]
    fits = normalizar_valores(fits) # normalizamos los valores para despues poder aplicar una suma o algo si hiciese falta
    fits, population = zip(*sorted(zip(fits, pop), key=lambda x:x[0])) # ordenamos por el primer valor, aunque no tiene mucho sentido ordenar ya que tendremos que comparar todos con todos y escoger primero los dominantes
    return fits, population

In [14]:


# recordemos como es un individuo ['182923', '500861', '299055', '152407', '30962', '313801', '394886', '335987' ...]
# el crossover puede ser un corte por n puntos
# donde el indice es el nodo i y el valor es el nodo elegido como al que "apunta", una primera mutacion podría ser que mute de nodo al que apunta
def crossover (ind1, ind2, pcross, n_puntos = 1):
    copia_in1 = copy.copy(ind1)
    copia_in2 = copy.copy(ind2)
    if random.random() < pcross:
      punto = len(ind1) // (n_puntos+1)
      for i in range(n_puntos):
        index_fin = punto*(i+2)
        index_inicio = punto*(i+1)
        if index_fin < len(ind1):
          aux = copia_in1[index_inicio:index_fin]
          copia_in1[index_inicio:index_fin] = copia_in2[index_inicio:index_fin]
          copia_in2[index_inicio:index_fin] = aux
        else:
          aux = copia_in1[index_inicio:]
          copia_in1[index_inicio:] = copia_in2[index_inicio:]
          copia_in2[index_inicio:] = aux
    return copia_in1, copia_in2
      
      
    
ind1, ind2 = crossover(pop[0],pop[1],1, 1)
print(f"ind1 antes: {pop[0][:len(pop[0])//2]} ")
print(f"ind2 antes: {pop[1][:len(pop[0])//2]}")
print(f"ind1 despues: {ind1[:len(pop[0])//2]}")
print(f"ind2 despues: {ind2[:len(pop[0])//2]}")


ind1 antes: ['75780', '317458', '521655', '496201', '190765', '379184', '394886', '335987', '198973', '154202', '176184', '9762', '366654', '460073', '16457', '216152', '546890', '126430', '306636', '475666', '114378', '276567', '55386', '374874', '165463', '229209', '141635', '104553', '289953', '27890', '448649', '64879', '59532', '374874', '518306', '127262', '186575', '84839', '532649', '214555', '216152', '172218', '270524', '296111', '256660', '280255', '326849', '186575', '435398', '399572', '482622', '29354', '359086', '108259', '59613', '208714', '49424', '342330', '318538', '34895', '502212', '30962', '361524', '123953', '301761', '19509', '59613', '57600', '217346', '127928', '30093', '79870', '137050', '321818', '321821', '377247', '500861', '370979', '313954', '285798', '154509', '330328', '546890', '546890', '190765', '185312', '379184', '543516', '141618', '45362', '149809', '313801', '198973', '383731', '141635', '447224', '275861', '279830', '14686', '403034', '340331'

In [18]:
print(conversiones)

{'75780': 0, '317458': 1, '186386': 2, '419862': 3, '176159': 4, '204831': 5, '104481': 6, '57383': 7, '401449': 8, '299055': 9, '176184': 10, '542780': 11, '366654': 12, '124992': 13, '16457': 14, '196682': 15, '546890': 16, '546891': 17, '34895': 18, '321619': 19, '112724': 20, '276567': 21, '55386': 22, '374874': 23, '55388': 24, '221281': 25, '432229': 26, '104553': 27, '383087': 28, '335987': 29, '448649': 30, '522378': 31, '59532': 32, '491667': 33, '518306': 34, '309411': 35, '182435': 36, '204966': 37, '532649': 38, '544939': 39, '542899': 40, '172218': 41, '270524': 42, '250046': 43, '204991': 44, '231620': 45, '497867': 46, '186575': 47, '346322': 48, '399572': 49, '276695': 50, '530647': 51, '18650': 52, '252125': 53, '59613': 54, '127199': 55, '379102': 56, '159972': 57, '366824': 58, '116969': 59, '202986': 60, '430314': 61, '350442': 62, '4328': 63, '458993': 64, '30962': 65, '35067': 66, '57600': 67, '217346': 68, '295177': 69, '49424': 70, '278': 71, '123159': 72, '3218

In [17]:
print(amazon_G.edges)

[('75780', '182923'), ('75780', '423864'), ('75780', '458993'), ('317458', '480479'), ('317458', '500861'), ('317458', '500865'), ('186386', '299055'), ('186386', '394703'), ('186386', '414657'), ('186386', '521655'), ('419862', '33251'), ('419862', '98081'), ('419862', '152407'), ('419862', '458993'), ('419862', '496201'), ('176159', '19509'), ('176159', '30962'), ('176159', '190765'), ('176159', '230788'), ('176159', '373756'), ('176159', '404948'), ('176159', '428737'), ('204831', '28513'), ('204831', '159972'), ('204831', '189286'), ('204831', '313801'), ('204831', '342330'), ('204831', '379184'), ('104481', '394886'), ('57383', '27890'), ('57383', '165463'), ('57383', '219781'), ('57383', '238491'), ('57383', '335987'), ('57383', '389895'), ('57383', '458993'), ('57383', '483076'), ('57383', '538129'), ('401449', '187924'), ('401449', '198973'), ('401449', '470620'), ('299055', '154202'), ('176184', '256445'), ('542780', '9762'), ('542780', '16457'), ('542780', '214555'), ('542780

In [31]:
def mutate (grafo, ind, pmut, transformation): # la mutación consistirá en cambiar un elemento por otro posible nodo con el que este conectado
    probs = [random.random() for prob in range(len(ind))]
    nuevo_elem = copy.copy(ind)
    adj_matrix = nx.to_numpy_array(grafo)
    for i, probability in enumerate(probs):
      if probability < pmut:
            
        nodo_vecino = random.choice(range(len(adj_matrix[i]))) #la matriz de adyacencia tiene los nodos vecinos, los nodos vecinos hay que convertirlos a numeros
        
        while adj_matrix[i][nodo_vecino] == 0 and nodo_vecino != i:
            nodo_vecino = random.choice(range(len(adj_matrix[nodo_vecino])))
        
        keys = list(transformation.keys())
        values = list(transformation.values())
        nuevo_elem[i] = keys[values.index(nodo_vecino)] # en el gen se pone el aleatorio de los vecinos pero transformado en el indice que le corresponde 
    return nuevo_elem
  
print(f"ind1 antes: {pop[0]} ")
print(f"ind1 despues: {mutate(amazon_G, pop[0], 0.2, conversiones)}")

ind1 antes: ['75780', '317458', '521655', '496201', '190765', '379184', '394886', '335987', '198973', '154202', '176184', '9762', '366654', '460073', '16457', '216152', '546890', '126430', '306636', '475666', '114378', '276567', '55386', '374874', '165463', '229209', '141635', '104553', '289953', '27890', '448649', '64879', '59532', '374874', '518306', '127262', '186575', '84839', '532649', '214555', '216152', '172218', '270524', '296111', '256660', '280255', '326849', '186575', '435398', '399572', '482622', '29354', '359086', '108259', '59613', '208714', '49424', '342330', '318538', '34895', '502212', '30962', '361524', '123953', '301761', '19509', '59613', '57600', '217346', '127928', '30093', '79870', '137050', '321818', '321821', '377247', '500861', '370979', '313954', '285798', '154509', '330328', '546890', '546890', '190765', '185312', '379184', '543516', '141618', '45362', '149809', '313801', '198973', '383731', '141635', '447224', '275861', '279830', '14686', '403034', '340331'

Ha mutado el indice 5 que es el nodo 204831 que tenia como valor 379184 y ahora tiene como valor 204831, es decir, antes se habia cogido la arista ('204831', '379184'), y ahora se ha cogido al mismo nodo. 
Tambien ha mutado el indice 8 que es el nodo 401449 que habia cogido la arista ('401449', '198973') y se ha cambiado por la arista ('401449', '470620')
Todo esto se ha sacado de amazon_G.edges